### PR TITLE
refactor(GoMod): Extract the `Graph` class and improve it a bit

### PR DESCRIPTION
**Pull request by fviernau**
The `GoMod` class was quite large and extracting the `Graph` class to a separate file removes a bit of distraction for the reader.

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/GoMod.kt [hunk 1]
```diff
@@ -338,9 +338,12 @@ class GoMod(
  * A class to represent a graph with dependencies. This representation is basically an adjacency list implemented by a
  * map whose keys are package identifiers and whose values are the identifiers of packages these packages depend on.
  */
-private class Graph(private val nodeMap: MutableMap<Identifier, Set<Identifier>> = mutableMapOf()) {
+private class Graph(nodeMap: MutableMap<Identifier, Set<Identifier>> = mutableMapOf()) {
     companion object : Logging
 
+    private val nodeMap: MutableMap<Identifier, MutableSet<Identifier>> =
+        nodeMap.mapValuesTo(mutableMapOf()) { it.value.toMutableSet() }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/GoMod.kt:345]: How about using `MutableSet<Identifier>` in the constructor instead? All but one caller already use a `MutableSet` anyway, so you only need to adjust the call in `breakCycles()` from

    return Graph(outgoingEdgesForNodes.mapValuesTo(mutableMapOf()) { it.value.toSet() })

to

    return Graph(outgoingEdgesForNodes.mapValuesTo(mutableMapOf()) { it.value.toMutableSet() })

File: analyzer/src/main/kotlin/managers/GoMod.kt [hunk 2]
```diff
@@ -515,6 +502,19 @@ private data class ModuleInfoFile(
     )
 }
 
+/**
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/GoMod.kt:393]: > Make the `Graph` agnostic of `Go`

Can you explain this a bit further in the commit message? At a first glance, there seems to be nothing Go-specific in that function, and the word "Go" isn't mentioned. My guess is that it's the *assumption* being made in the function that's Go-specific, i.e. that there's only one package without a version, and that that's the main package.

File: analyzer/src/main/kotlin/managers/GoMod.kt [hunk 3]
```diff
@@ -168,7 +168,7 @@ class GoMod(
             }
 
         val mainModuleId = moduleInfoForModuleName.values.single { it.main }.toId()
-        var graph = Graph(mutableMapOf(mainModuleId to emptySet()))
+        var graph = Graph().apply { addNode(mainModuleId) }
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/GoMod.kt:171]: Commit message: "an" -> "and"

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/GoMod.kt (hunk 2 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/GoMod.kt:393]: And even if the code is Go-specific, as the whole Graph still is Go-specific (Only used in Go-context, and that will probably not change, or?), why not just leave it there?

Comment (fviernau) [reply analyzer/src/main/kotlin/managers/GoMod.kt:393]: > why not just leave it there?

The file is very large. I find it easier to work with the code if this is in a separate file.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/GoMod.kt:393]: > If you mean the function: To me the logic belong more closely to `GoMod` than to that `Graph`. IMO a `Graph` is something everybody knows what it is about, if there are no built-in surprises like this.

Yes, I meant the function. As long as `Graph` is not useful / being used in another context than Go, I don't see a reason to extract it. But now that it already is, let's leave it like that.

**Review by sschuberth — APPROVED** [review]

**Comment by fviernau** [comment]
Merge regardless of the failing ClearlyDefined flaky test.

**Pull request merged**